### PR TITLE
feat(818): Evaluate Kata Containers over HyperD

### PR DESCRIPTION
**Pull request by parthasl**
## Context

Disk-based cache strategy implementation missing in executor-k8s.

## Objective

Add disk-based cache strategy implementation.

## References

[Build cache - disk storage strategy](https://github.com/screwdriver-cd/screwdriver/issues/1830)

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: config/pod.yaml.hbs [hunk 1]
```diff
@@ -49,16 +63,27 @@ spec:
       - name: dind-storage
         mountPath: /var/lib/docker
   {{/if}}
-
   initContainers:
   - name: launcher
     image: {{launcher_image}}
+    {{#if cache.diskEnabled}}
+    command: ['/bin/sh', '-c', 'chmod -R 777 /opt/sdpipelinecahe && chmod -R 777 /opt/sdjobcache && chmod -R 777 /opt/sdeventcache && cp -a /opt/sd/* /opt/launcher && cp -a /hab/* /opt/hab']
```
Comment (tkyi) [inline config/pod.yaml.hbs]:
```suggestion
    command: ['/bin/sh', '-c', 'chmod -R 777 /opt/sdpipelinecache && chmod -R 777 /opt/sdjobcache && chmod -R 777 /opt/sdeventcache && cp -a /opt/sd/* /opt/launcher && cp -a /hab/* /opt/hab']
```

File: config/pod.yaml.hbs [hunk 2]
```diff
@@ -49,16 +63,27 @@ spec:
...
+    {{else}}
     command: ['/bin/sh', '-c', 'cp -a /opt/sd/* /opt/launcher && cp -a /hab/* /opt/hab']
+    {{/if}}
     volumeMounts:
     - mountPath: /opt/launcher
       name: screwdriver
     - mountPath: /opt/hab
       name: habitat
+    {{#if cache.diskEnabled}}
+    - mountPath: /opt/sdpipelinecahe
```
Comment (tkyi) [inline config/pod.yaml.hbs]:
```suggestion
    - mountPath: /opt/sdpipelinecache
```

File: index.js [hunk 3]
```diff
@@ -124,33 +126,41 @@ class K8sExecutor extends Executor {
...
+     * @param  {Object} options.ecosystem                                  Screwdriver Ecosystem
+     * @param  {Object} options.ecosystem.api                              Routable URI to Screwdriver API
+     * @param  {Object} [options.ecosystem.pushgatewayUrl]                 Pushgateway URL for Prometheus
+     * @param  {Object} options.ecosystem.store                            Routable URI to Screwdriver Store
+     * @param  {Object} options.ecosystem.ui                               Routable URI to Screwdriver UI
+     * @param  {Object} options.kubernetes                                 Kubernetes configuration
+     * @param  {String} [options.kubernetes.token]                         API Token (loaded from /var/run/secrets/kubernetes.io/serviceaccount/token if not provided)
+     * @param  {String} [options.kubernetes.host=kubernetes.default]       Kubernetes hostname
+     * @param  {Number} [options.kubernetes.jobsNamespace=default]         Pods namespace for Screwdriver Jobs
+     * @param  {String} [options.kubernetes.baseImage]                     Base image for the pod
```
Comment (tkyi) [inline index.js]: is there a default value?

File: index.js [hunk 4]
```diff
@@ -124,33 +126,41 @@ class K8sExecutor extends Executor {
...
+     * @param  {Object} [options.kubernetes.nodeSelectors]                 Object representing node label-value pairs
+     * @param  {String} [options.launchVersion=stable]                     Launcher container version to use
+     * @param  {String} [options.prefix='']                                 Prefix for job name
+     * @param  {String} [options.fusebox]                                  Options for the circuit breaker (https://github.com/screwdriver-cd/circuit-fuses)
+     * @param  {Object} [options.requestretry]                             Options for the requestretry (https://github.com/FGRibreau/node-request-retry)
+     * @param  {Number} [options.requestretry.retryDelay]                  Value for retryDelay option of the requestretry
+     * @param  {Number} [options.requestretry.maxAttempts]                 Value for maxAttempts option of the requestretry
+     * @param  {String} [options.ecosystem.cache.strategy]                 Value for build cache - s3, disk
+     * @param  {String} [options.ecosystem.cache.path]                     Value for build cache path if options.cache.strategy is disk
```
Comment (tkyi) [inline index.js]: missing docs for `ecosystem.cache.compress`, `ecosystem.cache.md5check`

File: index.js [hunk 5]
```diff
@@ -124,33 +126,41 @@ class K8sExecutor extends Executor {
...
+     * @param  {Boolean} [options.kubernetes.dockerFeatureEnabled=false]   Whether to enable docker in docker on the executor k8 container
+     * @param  {Object} [options.kubernetes.nodeSelectors]                 Object representing node label-value pairs
+     * @param  {String} [options.launchVersion=stable]                     Launcher container version to use
+     * @param  {String} [options.prefix='']                                 Prefix for job name
+     * @param  {String} [options.fusebox]                                  Options for the circuit breaker (https://github.com/screwdriver-cd/circuit-fuses)
+     * @param  {Object} [options.requestretry]                             Options for the requestretry (https://github.com/FGRibreau/node-request-retry)
+     * @param  {Number} [options.requestretry.retryDelay]                  Value for retryDelay option of the requestretry
+     * @param  {Number} [options.requestretry.maxAttempts]                 Value for maxAttempts option of the requestretry
+     * @param  {String} [options.ecosystem.cache.strategy]                 Value for build cache - s3, disk
```
Comment (tkyi) [inline index.js]: can you put defaults in here too?

File: index.js [hunk 6]
```diff
@@ -250,29 +271,51 @@ class K8sExecutor extends Executor {
...
         const cpuValues = {
+            MAX: this.maxCpu,
             TURBO: this.turboCpu,
             HIGH: this.highCpu,
             LOW: this.lowCpu,
             MICRO: this.microCpu
         };
+
+        // PRs - set pipeline, job cache volume readonly and job cache dir to parent job cache dir
+        const regex = /^PR-([0-9]+)(?::[\w-]+)?$/gi;
```
Comment (tkyi) [inline index.js]: save as a constant?
might be good to have a more descriptive var name too

File: index.js [hunk 7]
```diff
@@ -250,29 +271,51 @@ class K8sExecutor extends Executor {
...
-     * @param  {String}   config.container  Container for the build to run in
-     * @param  {String}   config.token      JWT for the Build
+     * @param  {Object}   config                A configuration object
+     * @param  {Integer}  config.buildId        ID for the build
+     * @param  {Integer}  config.pipeline.id    pipelineId for the build
+     * @param  {Integer}  config.jobId          jobId for the build
+     * @param  {Integer}  config.eventId        eventId for the build
+     * @param  {String}   config.container      Container for the build to run in
+     * @param  {String}   config.token          JWT for the Build
+     * @param  {String}   config.jobName        jobName for the build
```
Comment (tkyi) [inline index.js]:
```suggestion
     * @param  {Integer}  config.buildId        ID for the build
     * @param  {Integer}  [config.pipeline.id]    pipelineId for the build
     * @param  {Integer}  [config.jobId]          jobId for the build
     * @param  {Integer}  config.eventId        eventId for the build
     * @param  {String}   config.container      Container for the build to run in
     * @param  {String}   config.token          JWT for the Build
     * @param  {String}   [config.jobName]        jobName for the build
```

File: index.js (hunk 3 above)
Comment (parthasl) [reply index.js]: followed executor-k8s-vm.

File: index.js (hunk 4 above)
Comment (parthasl) [reply index.js]: added

File: index.js (hunk 5 above)
Comment (parthasl) [reply index.js]: done

File: index.js (hunk 6 above)
Comment (parthasl) [reply index.js]: done

**Pull request merged**